blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/BlueTestResultFactory.java

```diff
@@ -0,0 +1,90 @@
+package io.jenkins.blueocean.service.embedded.rest;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import hudson.ExtensionPoint;
+import hudson.model.Run;
+import io.jenkins.blueocean.rest.Reachable;
+import io.jenkins.blueocean.rest.model.BlueTestResult;
+import io.jenkins.blueocean.rest.model.BlueTestSummary;
+import jenkins.model.Jenkins;
+
+public abstract class BlueTestResultFactory implements ExtensionPoint {
```

Comment:
@vivek big change to this PR... it turns out that the method of using `AbstractTestResultAction` to discover all the test cases breaks down on pretty much every plugin in the ecosystem and so many optional methods, many of them are unimplemented, just return empty lists for tests.

This factory extension point just returns a list of tests and automatically tallys a summary. Plugins would have to depend on this module in order to get tests displaying in Blue Ocean. There is an example implementation for JUnit in `BlueJUnitTestResult.FactoryImpl`.